File: gpd.m

function res = gpd(y,threshold )
[F,xi] = ecdf(y);
paramEsts = gpfit(y);
kHat      = paramEsts(1);
sigmaHat = paramEsts(2);

yi = gpcdf(xi,kHat,sigmaHat);
plot(xi,yi,'r-');
hold on;
stairs(xi,F,'b');
hold on;
probability = 1;
if xi(1) <=  threshold   &  threshold < xi( size(xi)  )
    for  i=2:size(xi)
    
     if (xi(i-1) <= threshold  )  & ( threshold  <  xi(i) ) 
        [k,b] = forkb(xi(i-1),yi(i-1),xi(i),yi(i))
        probability = k*threshold + b;
        break;
     end
    end
end
%plot(threshold,out,'o')
%hold on



res = [kHat,sigmaHat ,probability] ;
end

function [k,b] = forkb(x1,y1,x2,y2)
    k = (y2-y1)/(x2-x1);
    b = y1 - x1*k;

end

File: gpd.py
import matlab
import matlab.engine
import numpy as np
# engine = matlab.engine.start_matlab() # Start MATLAB process
# engine = matlab.engine.start_matlab("-desktop") # Start MATLAB process with graphic UI

class GPD:
    def __init__(self):
        self.engine = matlab.engine.start_matlab() # Start MATLAB process
        print("start GPD")



    def gpd(self,data,threshold):

        # 取得是过去slice的倍数的值  并且超过阈值
        slice = 1
        segment = len(data) // slice
        data = data[ -1*segment*slice: ]
        print(data)
        temp = []
        left = -1*segment*slice
        for i in range(segment) :
            right = left + slice
            if right == 0:
                mid = data[left:]
            else:
                mid = data[left:right]
            if max(mid) >= threshold:
                temp.append(max(mid))
            # print(mid)
            left += slice
        print("temp",temp)
        if not temp :
            return
        temp = matlab.double(temp)
        # temp = []
        threshold = [threshold]
        # threshold = threshold.tolist()
        threshold = matlab.double(threshold)
        # res  = self.engine.gpfit(temp)
        ans = self.engine.gpd(temp ,threshold)
        res = ans[0][0:2]
        probability = ans[0][2]
        print(probability)
        return res,probability



if __name__ == "__main__":
    aa = GPD()
    data = [i for i in range(20)]
    threshold = 1

    res = aa.gpd(data , threshold)

    # print(segment)
    print(res)